### PR TITLE
feat: hide caldav server settings if no app uses the caldav backend

### PR DESCRIPTION
**Pull request by st3iny**
* Resolves: _none_

## Summary

This PR introduces a dependency type to `info.xml` called `backends` which can be used by apps to express their intent to use a specific server-side backend. Currently, the only backend to be required is `caldav`.

The caldav server admin settings will be hidden if no app is requiring the `caldav` backend.

### Exemplary `info.xml`

```xml
<?xml version="1.0"?>
<info xmlns:xsi= "http://www.w3.org/2001/XMLSchema-instance"
	  xsi:noNamespaceSchemaLocation="https://apps.nextcloud.com/schema/apps/info.xsd">
	<id>calendar</id>
	<name>Calendar</name>
	<summary>A Calendar app for Nextcloud</summary>
	<version>4.8.0-alpha.1</version>
	<namespace>Calendar</namespace>
	<!-- [...] -->
	<dependencies>
		<php min-version="8.0" max-version="8.3" />
		<nextcloud min-version="27" max-version="30" />
		<backend>caldav</backend>
	</dependencies>
	<!-- [...] -->
</info>

```

### Hidden settings section

grafik

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by ChristophWurst** [comment]
> This PR introduces a new top level key in `info.xml` called `backends` which can be used by apps to express their intent to use a specific server-side backend. Currently, the only backend to be required is `caldav`.

If we are changing info.xsd we can also move this into the *dependencies* element. I think that makes the meaning clearer. The current approach is ambiguous as it could mean that the apps that list caldav as backend are *providing* it.

**Review by ChristophWurst — COMMENTED** [review]
Looks good so far!

File: resources/app-info.xsd [hunk 1]
```diff
@@ -633,6 +635,12 @@
         </xs:restriction>
     </xs:simpleType>
 
+    <xs:simpleType name="backend">
+        <xs:restriction base="xs:string">
+            <xs:enumeration value="caldav"/>
+        </xs:restriction>
+    </xs:simpleType>
+
```
Comment (ChristophWurst) [inline resources/app-info.xsd:643]: don't forget to also submit these changes to the appstore repo

**Review by ChristophWurst — APPROVED** [review]

File: resources/app-info.xsd (hunk 1 above)
Comment (st3iny) [reply resources/app-info.xsd]: https://github.com/nextcloud/appstore/pull/1414

**Comment by tcitworld** [comment]
> The caldav server admin settings will be hidden if no app is requiring the caldav backend.

So you won't be able to set the caldav settings if you want to use calendar sync with clients without the Calendar app? Apart from hiding the settings, is there any more consequences under consideration?

**Comment by ChristophWurst** [comment]
Yes, that is the behavior that was decided. We simply hide the settings. The case that someone uses the caldav backend without a Nextcloud frontend app is not covered.

**Comment by ChristophWurst** [comment]
I guess we should add a note to https://docs.nextcloud.com/server/latest/admin_manual/groupware/calendar.html for admins who can't find the settings

**Review by miaulalala — APPROVED** [review]

File: lib/public/App/IAppManager.php [hunk 2]
```diff
@@ -261,4 +266,14 @@ public function getDefaultApps(): array;
...
+
+	/**
+	 * Check whether the given backend is required by at least one app.
+	 *
+	 * @param self::BACKEND_* $backend Name of the backend, one of `self::BACKEND_*`
+	 * @return bool True if at least one app requires the backend
+	 *
+	 * @since 30.0.0
+	 */
+	public function isBackendRequired(string $backend): bool;
```
Comment (nickvergessen) [inline lib/public/App/IAppManager.php:278]: @st3iny Please document in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_30.html#id1
(Saw the ticket, just putting it to all PRs)

File: resources/app-info.xsd
Comment (nickvergessen) [inline resources/app-info.xsd:1]: It's a bit problematic that we don't have CI in place to confirm this, but https://github.com/nextcloud/appstore/blob/master/nextcloudappstore/api/v1/release/info.xsd needs to be kept in sync.

Comment (ChristophWurst) [reply resources/app-info.xsd:1]: https://github.com/nextcloud/appstore/pull/1414

**Pull request merged**